binary_transparency/firmware/cmd/ft_witness/internal/http/witness.go

```diff
@@ -0,0 +1,155 @@
+func NewWitness(ws WitnessStore, logURL string, pollInterval time.Duration) *Witness {
+	return &Witness{
+		ws:           ws,
+		logURL:       logURL,
+		pollInterval: pollInterval,
+	}
+}
+
+// getCheckpoint returns a checkpoint which is registered with witness
+func (s *Witness) getCheckpoint(w http.ResponseWriter, r *http.Request) {
+	checkpoint, err := s.ws.RetrieveCP()
+	if err != nil {
+		http.Error(w, err.Error(), httpStatusForErr(err))
+		return
+	}
+
+	js, err := json.Marshal(checkpoint)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(js)
+}
+
+// httpStatusForErr maps status codes to HTTP errors.
+func httpStatusForErr(e error) int {
+	switch status.Code(e) {
+	case codes.OK:
+		return http.StatusOK
+	case codes.NotFound:
+		return http.StatusNotFound
+	default:
+		return http.StatusInternalServerError
+	}
+	// unreachable
+}
+
+// RegisterHandlers registers HTTP handlers for firmware transparency endpoints.
+func (s *Witness) RegisterHandlers(r *mux.Router) {
+	r.HandleFunc(fmt.Sprintf("/%s", api.WitnessGetCheckpoint), s.getCheckpoint).Methods("GET")
+}
+
+// Poll periodically polls the FT log for updating the witness checkpoint.
+func (s *Witness) Poll(ctx context.Context) error {
+
+	ticker := time.NewTicker(s.pollInterval)
+	ftURL, err := url.Parse(s.logURL)
+	if err != nil {
+		return fmt.Errorf("failed to parse FT log URL: %w", err)
+	}
+	glog.Infof("Polling FT log %q...", ftURL)
+	c := client.ReadonlyClient{LogURL: ftURL}
+	lv := verify.NewLogVerifier()
+	for {
+
+		wcp, err := s.ws.RetrieveCP()
+		if err != nil {
+			glog.Warningf("Failed to retrieve store logcheckpoint: %q", err)
+			continue
+		}
+
+		select {
+		case <-ticker.C:
+			//
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+		cp, err := c.GetCheckpoint()
+		if err != nil {
+			glog.Warningf("Failed to update LogCheckpoint: %q", err)
+			continue
+		}
+		if cp.TreeSize <= wcp.TreeSize {
+			continue
+		}
+		glog.V(1).Infof("Got newer checkpoint %s", cp)
+		// Perform consistency check only for non-zero saved witness tree size
+		if wcp.TreeSize != 0 {
+			consistency, err := c.GetConsistencyProof(api.GetConsistencyRequest{From: wcp.TreeSize, To: cp.TreeSize})
+			if err != nil {
+				glog.Warningf("Failed to fetch the Consistency: %q", err)
+				continue
+			}
+			glog.V(1).Infof("Printing the latest Consistency Proof Information")
+			glog.V(1).Infof("Consistency Proof = %x", consistency.Proof)
+
+			//Verify the fetched consistency proof
+			if err := lv.VerifyConsistencyProof(int64(wcp.TreeSize), int64(cp.TreeSize), wcp.RootHash, cp.RootHash, consistency.Proof); err != nil {
+				// Verification of Consistency Proof failed!!
+				glog.Warningf("Failed verification of Consistency proof %q", err)
+				continue
+			}
+			glog.V(1).Infof("Consistency proof for Treesize %d verified", cp.TreeSize)
+		}
+		wcp = *cp
+
+		if s.ws.StoreCP(wcp) != nil {
```

Comment:
```suggestion
		if s.ws.StoreCP(*cp) != nil {
```

I'm not sure the variable juggling makes this any more readable.

Reply:
DONE!